llvm/lib/Target/SystemZ/SystemZISelLowering.cpp

```diff
@@ -8701,95 +8734,341 @@ SDValue SystemZTargetLowering::combineSETCC(
+    const auto &&Op1SDVals = simplifyAssumingCCVal(Op1, Op0CC, DCI);
+    if (Op0SDVals.empty() || Op1SDVals.empty())
       return false;
-    if (CompareRHS->getAPIntValue() == FalseVal->getAPIntValue())
-      Invert = !Invert;
-    else if (CompareRHS->getAPIntValue() != TrueVal->getAPIntValue())
+    auto &&Op0CCVals = getConstFromConstSDVals(Op0SDVals);
+    const auto &&Op1CCVals = getConstFromConstSDVals(Op1SDVals);
+    if (Op0CCVals.empty() || Op1CCVals.empty())
       return false;
-
-    // Compute the effective CC mask for the new branch or select.
-    auto *NewCCValid = dyn_cast<ConstantSDNode>(CompareLHS->getOperand(2));
-    auto *NewCCMask = dyn_cast<ConstantSDNode>(CompareLHS->getOperand(3));
-    if (!NewCCValid || !NewCCMask)
+    std::transform(Op0CCVals.begin(), Op0CCVals.end(), Op1CCVals.begin(),
+                   Op0CCVals.begin(), emulateTMCCMask);
+    if (std::any_of(Op0CCVals.begin(), Op0CCVals.end(),
+                    [](const auto &CC) { return CC < 0 || CC > 3; }))
       return false;
-    CCValid = NewCCValid->getZExtValue();
-    CCMask = NewCCMask->getZExtValue();
-    if (Invert)
-      CCMask ^= CCValid;
-
-    // Return the updated CCReg link.
-    CCReg = CompareLHS->getOperand(4);
+    int NewCCMask = 0;
+    for (auto CC : Op0CCVals) {
+      NewCCMask <<= 1;
+      NewCCMask |= (CCMask & (1 << (3 - CC))) != 0;
+    }
+    CCReg = Op0CC;
+    CCMask = NewCCMask;
     return true;
   }
+  if (CCNode->getOpcode() != SystemZISD::ICMP ||
+      CCValid != SystemZ::CCMASK_ICMP)
+    return false;
 
-  // Optimize the case where CompareRHS is (SRA (SHL (IPM))).
-  if (CompareLHS->getOpcode() == ISD::SRA) {
-    auto *SRACount = dyn_cast<ConstantSDNode>(CompareLHS->getOperand(1));
-    if (!SRACount || SRACount->getZExtValue() != 30)
-      return false;
-    auto *SHL = CompareLHS->getOperand(0).getNode();
-    if (SHL->getOpcode() != ISD::SHL)
-      return false;
-    auto *SHLCount = dyn_cast<ConstantSDNode>(SHL->getOperand(1));
-    if (!SHLCount || SHLCount->getZExtValue() != 30 - SystemZ::IPM_CC)
-      return false;
-    auto *IPM = SHL->getOperand(0).getNode();
-    if (IPM->getOpcode() != SystemZISD::IPM)
-      return false;
-
-    // Avoid introducing CC spills (because SRA would clobber CC).
-    if (!CompareLHS->hasOneUse())
+  SDValue CmpOp0 = CCNode->getOperand(0);
+  SDValue CmpOp1 = CCNode->getOperand(1);
+  SDValue CmpOp2 = CCNode->getOperand(2);
+  auto [Op0CC, Op0CCValid] = findCCUse(CmpOp0);
+  if (Op0CC != SDValue()) {
+    const auto &&Op0SDVals = simplifyAssumingCCVal(CmpOp0, Op0CC, DCI);
+    const auto &&Op1SDVals = simplifyAssumingCCVal(CmpOp1, Op0CC, DCI);
+    if (Op0SDVals.empty() || Op1SDVals.empty())
       return false;
-    // Verify that the ICMP compares against zero.
-    if (CompareRHS->getZExtValue() != 0)
+    auto &&Op0CCVals = getConstFromConstSDVals(Op0SDVals);
+    const auto &&Op1CCVals = getConstFromConstSDVals(Op1SDVals);
```

Comment:
Again with the `int` truncation ... this also should operate on the full `APInt`s.